### DISCUSSION TITLE
Grid Peak Shaving - Toggle

### DISCUSSION THREAD

**Discussion by Hatios** [kellerza/sunsynk#278] [General]
Hi Kellerza,

Your work on this addon is much appreciated, everything works so well with HA to automate my parallel Sunsynk single-phase setup.

The only thing causing me grief is being able to toggle Grid Peak Shaving on and off.
I have added the custom sensor on Modbus register 293 to enable changing the wattage of Grid Peak shaving, but I've not been able to figure out how to do what's required to enable the toggle.

I understand it is on Modbus register 280 but this is a multi-bit register.
According to the doc it is bits 8-11 responsible for Grid Peak Shaving, with 0 being disabled and 1 being enabled.
I tried to understand how the timer is formatted for similar bitmask requirements but after a couple of hours of trial and error, I wasn't successful.

This causes an issue as the Sunsunk firmware won't export from the battery when Grid Peak Shaving is toggled on, regardless of the selling mode.

If you are able to help, or better yet add this function to the next release, that would be great!

Thanks

**Comment by kellerza** [comment]
These are three-phase definitions, but using the correct register & adapting the bits it might help you get to the answer

Once you have a working sensor, a PR would be appreciated

`0b11` is binary for 3 and `<<` is left shift (*2). The best place to experiment with this is a Python REPL (just run python)

```python
    SelectRWSensor(
        178,
        "Gen peak-shaving",
        options={0b10 << 2: "Disable", 0b11 << 2: "Enable"},
        bitmask=0b11 << 2,
    ),
    SelectRWSensor(
        178,
        "Grid peak-shaving",
        options={0b10 << 4: "Disable", 0b11 << 4: "Enable"},
        bitmask=0b11 << 4,
    ),
```

**Reply by Hatios** [reply]
Thanks for coming back to me.

I don't have an IDE setup and unfortunately very little experience with Python, but I'm familiar with editing the mysensors.py file for the addon so I gave this a go on there.
See my example below.
![image](https://github.com/kellerza/sunsynk/assets/147992465/efc0823b-39a0-4f71-b033-6d6afa8599d4)


From your above example, is sensor 178 not used for Load Power on the single phase definitions?
I tried using the above like so in the mysensors.py file based on your advice, but unfortunately using 2 or 4 for the shift, or trying sensor 178 or 280 both still yield errors like the below:
_09:06:20 WARNING Cannot publish Grid Peak Shaving: value is None
09:06:25 WARNING grid_peak_shaving: Unknown register value 16
09:06:25 WARNING grid_peak_shaving: Unknown register value 16
09:06:31 WARNING grid_peak_shaving: Unknown register value 0
09:06:31 WARNING grid_peak_shaving: Unknown register value 0_

For some context, my concern is that my 2 inverters in parallel can draw from the grid at more than the current their MCB is rated for, for quite some time before tripping which causes significant overheating in the consumer unit. I fix this with Grid Peak Shaving which works perfectly, but then all export just stops till that is disabled.
This hasn't been a big problem with the amount of sun we've been having but there was one day where the system topped up the batteries from grid and with other loads in the house (my whole home is on UPS side), the inverters were drawing around 80 amps when this should be limited to 60!
There's the potential for me being away and Grid Peak Shaving being left "off" resulting in damage or tripping a breaker while I'm not there to reset it. I want to automate it being toggled on during grid charging and off during grid export.

Thanks!

**Reply by kellerza** [reply]
You have to change the 178 to 280. 178 is for three-phase.

Apologies if my first sentence was not clear enough

**Reply by kellerza** [reply]
You might also have to adapt the actual bits / shift operation to what you see in the modbus register documentation.

**Comment by Hatios** [comment]
I believe I have this working, without any shifting.
I looked at the error I was getting when just trying the below (not making any changed just in case it sent a weird command but looking at the values in the log):
SENSORS += SelectRWSensor(280, "Grid Peak Shaving", options={0: "Disable", 1: "Enable"},),

This reported back that it unknown value 8192, which is a single bit in the 1 position in the 14th slot which i believe is something to do with "On Grid Always On", not sure what that is but I will leave it alone :D

I then calculated the decimal value of the 14 digits if position 9 was in the on position as 8488.
I updated my sensor line to use these numbers instead of 0 and 1.

SENSORS += SelectRWSensor(280, "Grid Peak Shaving", options={8192: "Disable", 8448: "Enable"},),

Hey presto, now when I toggle the setting on the sunsynk app, it is read correctly in HA and if i toggle it, it works on the inverter too!

I appreciate that this setting is adjusting the entire 16 bit register, but I do not believe I have need to touch any of the other settings that this block seems to contain.
![image](https://github.com/kellerza/sunsynk/assets/147992465/4a7a8b8c-b7c0-4913-87b0-5e2ae65be5a7)

**Reply by kellerza** [reply]
I added some grid peak shaving settings, see <https://github.com/kellerza/sunsynk/blob/main/src/sunsynk/definitions/single_phase.py#L245-L251>

Would appreciate feedback on what works, what not

**Reply by RenierBoshoff** [reply]
I am going to have a look at this.
Will investigate how to show it on the dashboard.

**Reply by RenierBoshoff** [reply]
Do I need to add the code manually or is it included in the Options > Inverter Driver > pymodbus of the Sunsynk/Deye Inverter Add-on (multi)?

**Reply by RenierBoshoff** [reply]
> Do I need to add the code manually or is it included in the Options > Inverter Driver > pymodbus of the Sunsynk/Deye Inverter Add-on (multi)?

I will read here: https://kellerza.github.io/sunsynk/reference/mysensors

**Comment by RenierBoshoff** [comment]
Hallo @kellerza 

Apologies but I am not sure what I am missing.

Here is the mysensors.py


from sunsynk import AMPS, CELSIUS, KWH, VOLT, WATT, SensorDefinitions
from sunsynk.rwsensors import NumberRWSensor, SelectRWSensor, TimeRWSensor, SwitchRWSensor
from sunsynk.sensors import MathSensor, TempSensor

# Initialize the sensor definitions
SENSORS = SensorDefinitions()

# Add your custom sensors
SENSORS += (SwitchRWSensor(280, "Peak shaving", bitmask=0x0F, on=1),)      # ?
SENSORS += (SwitchRWSensor(280, "Gen peak shaving", bitmask=0xF0, on=0x10),)
SENSORS += (SwitchRWSensor(280, "Grid peak shaving", bitmask=0xF00, on=0x100),)
SENSORS += (SwitchRWSensor(280, "Grid always on", bitmask=0xF000, on=0x1000),)
SENSORS += (NumberRWSensor(292, "Gen peak shaving power", WATT, 1, min=0, max=RATED_POWER),)
SENSORS += (NumberRWSensor(293, "Grid peak shaving power", WATT, 1, min=0, max=RATED_POWER),)
SENSORS += (SwitchRWSensor(243, "Priority Load", bitmask=1),)



However i get these problems in the addon log:

19:42:45 INFO    Using Single phase sensor definitions.
19:42:45 INFO    Importing /share/hass-addon-sunsynk/mysensors.py...
Traceback (most recent call last):
  File "/usr/local/lib/python3.12/site-packages/ha_addon_sunsynk_multi/helpers.py", line 47, in import_mysensors
    mod = import_module("mysensors", str(root))
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/site-packages/ha_addon_sunsynk_multi/helpers.py", line 20, in import_module
    mod_obj = _import_module(mod_name)
              ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 999, in exec_module
  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
  File "/share/hass-addon-sunsynk/mysensors.py", line 1, in <module>
    from sunsynk import AMPS, CELSIUS, KWH, VOLT, WATT, SensorDefinitions
ImportError: cannot import name 'SensorDefinitions' from 'sunsynk' (/usr/local/lib/python3.12/site-packages/sunsynk/__init__.py)
19:42:45 ERROR   Error importing /share/hass-addon-sunsynk/mysensors.py: cannot import name 'SensorDefinitions' from 'sunsynk' (/usr/local/lib/python3.12/site-packages/sunsynk/__init__.py)
19:42:45 ERROR   Unknown sensor specified: peak_shaving
19:42:45 ERROR   Unknown sensor specified: gen_peak_shaving
19:42:45 ERROR   Unknown sensor specified: grid_peak_shaving_power
19:42:45 INFO    Added hidden sensors as other sensors depend on it: Device type, Protocol, Serial, Rated power
19:42:45 INFO    Connecting to /dev/ttyUSB0
19:42:45 INFO    PyModbus 3.8.3 Serial: /dev/ttyUSB0
19:42:45 INFO    Reading startup sensors Device type, Protocol, Serial
19:42:45 INFO    ############################################################
19:42:45 INFO                 Inverter serial number '****0152'


Any pointers would be appreciated!

**Comment by RenierBoshoff** [comment]
Satisfaction!!

<img width="529" height="669" alt="image" src="https://github.com/user-attachments/assets/00b0b216-d167-47c5-aac4-4993630b9e80" />

**Reply by RenierBoshoff** [reply]
# mysensors.py 

# Define or import your rated power
RATED_POWER = 5000

from sunsynk import AMPS, CELSIUS, KWH, VOLT, WATT
from sunsynk.rwsensors import NumberRWSensor, SelectRWSensor, TimeRWSensor, SwitchRWSensor

from sunsynk.rwsensors import (
	NumberRWSensor,
	SelectRWSensor,
	SwitchRWSensor,
	SystemTimeRWSensor,
	TimeRWSensor,
)
from sunsynk.sensors import (
	MathSensor,
	Sensor,
	SensorDefinitions,
	TempSensor,
)


# Initialize the sensor definitions
SENSORS = SensorDefinitions()

# Extend with your own Sunsynk sensors
SENSORS += (
    SwitchRWSensor(280, "Peak shaving",       bitmask=0x0F,   on=0x01),
    SwitchRWSensor(280, "Gen peak shaving",   bitmask=0xF0,   on=0x10),
    SwitchRWSensor(280, "Grid peak shaving",  bitmask=0xF00,  on=0x100),
    SwitchRWSensor(280, "Grid always on",     bitmask=0xF000, on=0x1000),
    NumberRWSensor(292, "Gen peak shaving power",  WATT, 1, min=0, max=RATED_POWER),
    NumberRWSensor(293, "Grid peak shaving power", WATT, 1, min=0, max=RATED_POWER),
    NumberRWSensor(230, "Grid Charge Battery current", AMPS, min=0, max=185),
    NumberRWSensor(210, "Battery Max Charge current", AMPS, min=0, max=185),
    NumberRWSensor(211, "Battery Max Discharge current", AMPS, min=0, max=185),
    
)

**Reply by maslyankov** [reply]
@RenierBoshoff can you do a PR with these definitions? If you need help, I can help if needed. Adding those for three-phase inverters would be useful as well. Good stuff.

**Reply by RenierBoshoff** [reply]
> @RenierBoshoff can you do a PR with these definitions? If you need help, I can help if needed. Adding those for three-phase inverters would be useful as well. Good stuff.

Hallo @maslyankov , sure thing. Can I read up how to do it so it will be useful?
And can you perhaps remind me how to get the nice formatting?

Thanks!

**Reply by RenierBoshoff** [reply]
Reading here... https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests
